cluster-autoscaler/core/scale_down.go

```diff
@@ -266,55 +268,23 @@ func (sd *ScaleDown) TryToScaleDown(nodes []*apiv1.Node, pods []*apiv1.Pod, pdbs
 	// to recreate on other nodes.
 	emptyNodes := getEmptyNodes(candidates, pods, sd.context.MaxEmptyBulkDelete, sd.context.CloudProvider)
 	if len(emptyNodes) > 0 {
+		nodeDeletionStart := time.Now()
 		confirmation := make(chan errors.AutoscalerError, len(emptyNodes))
-		for _, node := range emptyNodes {
-			glog.V(0).Infof("Scale-down: removing empty node %s", node.Name)
-			sd.context.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaleDownEmpty", "Scale-down: removing empty node %s", node.Name)
-			simulator.RemoveNodeFromTracker(sd.usageTracker, node.Name, sd.unneededNodes)
-			go func(nodeToDelete *apiv1.Node) {
-
-				deleteErr := deleteNodeFromCloudProvider(nodeToDelete, sd.context.CloudProvider,
-					sd.context.Recorder, sd.context.ClusterStateRegistry)
-				if deleteErr == nil {
-					if readinessMap[nodeToDelete.Name] {
-						metrics.RegisterScaleDown(1, metrics.Empty)
-					} else {
-						metrics.RegisterScaleDown(1, metrics.Unready)
-					}
-				}
-				confirmation <- deleteErr
-			}(node)
-		}
-		var finalError errors.AutoscalerError
-
-		startTime := time.Now()
-		for range emptyNodes {
-			timeElapsed := time.Now().Sub(startTime)
-			timeLeft := MaxCloudProviderNodeDeletionTime - timeElapsed
-			if timeLeft < 0 {
-				finalError = errors.NewAutoscalerError(errors.TransientError, "Failed to delete nodes in time")
-				break
-			}
-			select {
-			case err := <-confirmation:
-				if err != nil {
-					glog.Errorf("Problem with empty node deletion: %v", err)
-					finalError = err
-				}
-			case <-time.After(timeLeft):
-				finalError = errors.NewAutoscalerError(errors.TransientError, "Failed to delete nodes in time")
-			}
-		}
-		if finalError == nil {
+		sd.scheduleDeleteEmptyNodes(emptyNodes, readinessMap, confirmation)
+		err := waitForEmptyNodesDeleted(emptyNodes, confirmation)
```

Comment:
I don't like the fact that scheduleDeleteEmptyNodes is a method and waitForEmptyNodesDeleted is a function. I understand it doesn't need to access any fields on sd, but still I would make it a method for consistency.